spec/assign-extended.test.sh

```diff
@@ -99,6 +99,366 @@ echo $?
 127
 ## END
 
+#### declare
+test_var1=111
+readonly test_var2=222
+export test_var3=333
+declare -n test_var4=test_var1
+f1() {
+  local test_var5=555
+  {
+    echo '[declare]'
+    declare
+    echo '[readonly]'
+    readonly
+    echo '[export]'
+    export
+    echo '[local]'
+    local
+  } | grep -E '^\[|^\b.*test_var.\b'
+}
+f1
+## STDOUT:
+[declare]
+test_var1='111'
+test_var2='222'
+test_var3='333'
+test_var4='test_var1'
+test_var5='555'
+[readonly]
+declare -r test_var2='222'
+[export]
+declare -x test_var3='333'
+[local]
+test_var5='555'
+## END
+## OK bash STDOUT:
```

Comment:
Maybe instead of listing exact output of OSH and bash, you can `eval` the output of `declare -p`?

I think that's the property you want, right?

It makes the test a little more meaningful if both bash and Oil pass the exact same way.

-----

Not all tests have to be like that , but I think at least one should no?

Reply:
> Maybe instead of listing exact output of OSH and bash, you can eval the output of declare -p?

Thank you for the suggestion! That's a good idea! I added such a test d1c85b43, but it is still a very simple test.

Actually, the problem is "*how to check the results*" of `eval` & `declare`: The existence, the flags, and the values of each variable should be checked (*i.e.*, printed for the test). In the case of indexed/associative array variables, the existence and the value of each element, and *non-existence* of redundant elements should also be checked (*i.e.*, printed). If one wants to test everything in the exact same way in Bash and Oil, it seems that one needs to implement another `declare -p` in pure shell scripts (which may involve other bugs/divergences on the detailed behavior of each shell), so I gave up.

> I think that's the property you want, right?

Yes! Thank you.

Reply:
Yes that's true, it's tricky.  One possibility is to `sed` to different variable names, and then check for equality.  Something like:

```
eval -- $(declare -p | grep ^myarray | sed s/^myarray=/restored=/)
# now check myarray == restored

arrays_equal ...   # I wrote a function that takes two lengths and then @A and @B concatenated
```

But yes this is a lot of trouble ...

Actually in Oil we may be able to help with this.  We simply have

```
if (A == B) {  # are string, arrays, associative arrays equal?
}
```

but that's in the future